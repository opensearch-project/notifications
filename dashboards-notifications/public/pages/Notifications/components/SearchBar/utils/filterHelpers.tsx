/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiComboBox, EuiFieldText, EuiFormRow, EuiSpacer } from '@elastic/eui';
import React from 'react';
import {
  CHANNEL_TYPE,
  NOTIFICATION_SOURCE,
  SEVERITY_TYPE,
} from '../../../../../utils/constants';
import { FilterType } from '../Filter/Filters';

export interface FilterParamsType {
  'status_list.config_name'?: string[];
  'status_list.config_type'?: Array<keyof typeof CHANNEL_TYPE>;
  'event_source.feature'?: Array<keyof typeof NOTIFICATION_SOURCE>;
  'event_source.severity'?: Array<keyof typeof SEVERITY_TYPE>;
  'status_list.delivery_status.status_code'?: string[];
}

export type FilterFieldType =
  | 'Channel'
  | 'Channel type'
  | 'Severity'
  | 'Source'
  | 'Status';

export type FilterOperatorType =
  | 'is'
  | 'is not'
  | 'is one of'
  | 'is not one of';

export type FilterValueType = string | number | string[] | number[];

export const getFilterFieldOptions = () => {
  return [
    { label: 'Channel' },
    { label: 'Channel type' },
    { label: 'Severity' },
    { label: 'Source' },
    { label: 'Status' },
  ];
};

export const getFilterOperatorOptions = (field: FilterFieldType) => {
  if (field === 'Channel' || field === 'Status') {
    return [
      { label: 'is' },
      // { label: 'is not' },
    ];
  }
  return [
    { label: 'is' },
    // { label: 'is not' },
    { label: 'is one of' },
    // { label: 'is not one of' },
  ];
};

export const getOperatorString = (
  operator: FilterOperatorType,
  inverted: boolean
) => {
  if (operator === 'is' || operator === 'is not')
    return inverted ? 'is not' : 'is';
  if (operator === 'is one of' || operator === 'is not one of')
    return inverted ? 'is not one of' : 'is one of';
  return '';
};

export const getValueComponent = (
  field: FilterFieldType,
  operator: FilterOperatorType,
  value: any,
  setValue: (v: any) => void,
  availableChannels: { [feature: string]: string }
) => {
  // only Channel filters should be inputbox, others should be combobox
  if (field === 'Channel')
    return (
      <>
        <EuiSpacer size="s" />
        <EuiFormRow label="Value">
          <EuiFieldText
            placeholder="Enter a value"
            value={value || ''}
            onChange={(e) => setValue(e.target.value)}
          />
        </EuiFormRow>
      </>
    );

  let options;

  if (field === 'Channel type')
    options = Object.entries(availableChannels).map(([key, value]) => ({
      label: value,
      value: key,
    }));
  else if (field === 'Severity')
    options = Object.entries(SEVERITY_TYPE).map(([key, value]) => ({
      label: value,
      value: key,
    }));
  else if (field === 'Source')
    options = Object.entries(NOTIFICATION_SOURCE).map(([key, value]) => ({
      label: value,
      value: key,
    }));
  else if (field === 'Status')
    options = [{ label: 'Sent' }, { label: 'Error' }];

  const singleSelection = isSingleSelection(operator);
  const placeholder = singleSelection ? 'Select a value' : 'Select values';

  return (
    <>
      <EuiSpacer size="s" />
      <EuiFormRow label="Value">
        <EuiComboBox
          fullWidth
          placeholder={placeholder}
          options={options}
          selectedOptions={value || []}
          singleSelection={singleSelection}
          onChange={setValue}
          isClearable={true}
        />
      </EuiFormRow>
    </>
  );
};

export const isSingleSelection = (filterOrOperator: FilterType | string) => {
  const checkSingleSelection = (operator: string) =>
    operator === 'is' || operator === 'is not';

  return checkSingleSelection(
    typeof filterOrOperator === 'string'
      ? filterOrOperator
      : filterOrOperator.operator
  );
};

const toBackendField = (field: FilterFieldType) => {
  switch (field) {
    case 'Channel':
      return 'status_list.config_name';
    case 'Channel type':
      return 'status_list.config_type';
    case 'Severity':
      return 'event_source.severity';
    case 'Source':
      return 'event_source.feature';
    case 'Status':
      return 'status_list.delivery_status.status_code';
    default:
      return '';
  }
};

export const filtersToQueryParams = (filters: FilterType[]) => {
  const filterParams: FilterParamsType = {};
  filters
    .filter((filter) => !filter.disabled && filter.value != null)
    .forEach((filter) => {
      const field = toBackendField(filter.field);
      if (!field) return;
      if (!filterParams[field]) filterParams[field] = [];

      if (field === 'status_list.delivery_status.status_code') {
        const queryForSuccess =
          (filter.value === 'Sent' && !filter.inverted) ||
          (filter.value === 'Error' && filter.inverted);
        filterParams[field]!.push(queryForSuccess ? '200' : '!200');
        return;
      }

      const inverted = filter.inverted ? '!' : '';
      const values =
        typeof filter.value === 'string'
          ? [filter.value]
          : filter.value!.map((option) => option.value || option.label);
      for (const value of values) {
        filterParams[field]!.push(`${inverted}${value}`);
      }
    });
  return filterParams;
};
