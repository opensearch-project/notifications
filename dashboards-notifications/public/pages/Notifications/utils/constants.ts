/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const DEFAULT_PAGE_SIZE_OPTIONS = [5, 10, 20, 50];
