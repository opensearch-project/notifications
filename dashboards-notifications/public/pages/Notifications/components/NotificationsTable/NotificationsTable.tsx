/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiBasicTable,
  EuiHealth,
  EuiLink,
  EuiTableFieldDataColumnType,
  EuiTableSortingType,
} from '@elastic/eui';
import { Criteria } from '@elastic/eui/src/components/basic_table/basic_table';
import { Pagination } from '@elastic/eui/src/components/basic_table/pagination_bar';
import _ from 'lodash';
import React from 'react';
import {
  ChannelStatus,
  NotificationItem,
} from '../../../../../models/interfaces';
import { ContentPanel } from '../../../../components/ContentPanel';
import { ModalConsumer } from '../../../../components/Modal';
import {
  CHANNEL_TYPE,
  NOTIFICATION_SOURCE,
  SEVERITY_TYPE,
} from '../../../../utils/constants';
import { renderTime } from '../../../../utils/helpers';
import { TableFlyout } from './Flyout/TableFlyout';

interface NotificationsTableProps {
  items: NotificationItem[];
  onTableChange: ({
    page: tablePage,
    sort,
  }: Criteria<NotificationItem>) => void;
  pagination: Pagination;
  sorting: EuiTableSortingType<NotificationItem>;
  loading: boolean;
}

export function NotificationsTable(props: NotificationsTableProps) {
  const columns: EuiTableFieldDataColumnType<NotificationItem>[] = [
    {
      field: 'event_source.title',
      name: 'Notification',
      sortable: true,
      truncateText: true,
      render: (title, item) => (
        <ModalConsumer>
          {({ onShow }) => (
            <EuiLink
              onClick={() => onShow(TableFlyout, { notificationItem: item })}
            >
              {title}
            </EuiLink>
          )}
        </ModalConsumer>
      ),
    },
    {
      field: 'event_source.feature',
      name: 'Source type',
      sortable: true,
      truncateText: true,
      render: (source) => _.get(NOTIFICATION_SOURCE, source, '-'),
    },
    {
      field: 'event_source.severity',
      name: 'Severity',
      sortable: true,
      truncateText: false,
      render: (severity) => _.get(SEVERITY_TYPE, severity, '-'),
    },
    {
      field: 'last_updated_time_ms',
      name: 'Time sent',
      sortable: true,
      truncateText: false,
      render: renderTime,
      dataType: 'date',
    },
    {
      field: 'success',
      name: 'Sent status',
      sortable: false,
      render: (success, item: NotificationItem) => {
        const color = success ? 'success' : 'danger';
        const label = success ? 'Sent' : 'Error';
        return (
          <EuiHealth color={color}>
            {success ? (
              label
            ) : (
              <ModalConsumer>
                {({ onShow }) => (
                  <EuiLink
                    onClick={() =>
                      onShow(TableFlyout, { notificationItem: item })
                    }
                  >
                    {label}
                  </EuiLink>
                )}
              </ModalConsumer>
            )}
          </EuiHealth>
        );
      },
    },
    {
      field: 'status_list',
      name: 'Channels',
      sortable: false,
      truncateText: true,
      render: (status_list: ChannelStatus[]) =>
        status_list.length === 1
          ? status_list[0].config_name
          : `${status_list.length} channels`,
    },
    {
      field: 'status_list',
      name: 'Channel types',
      sortable: false,
      truncateText: false,
      render: (status_list: ChannelStatus[]) =>
        status_list
          .map((channel) => _.get(CHANNEL_TYPE, channel.config_type, '-'))
          .join(', '),
    },
  ];

  return (
    <>
      <ContentPanel
        bodyStyles={{ padding: 'initial' }}
        title="Notification history"
        titleSize="m"
      >
        <EuiBasicTable
          columns={columns}
          tableLayout="auto"
          itemId="id"
          isSelectable={true}
          items={props.items}
          onChange={props.onTableChange}
          pagination={props.pagination}
          sorting={props.sorting}
          loading={props.loading}
        />
      </ContentPanel>
    </>
  );
}
