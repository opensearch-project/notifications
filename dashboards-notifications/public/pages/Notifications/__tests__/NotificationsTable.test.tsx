/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { EuiTableSortingType } from '@elastic/eui';
import { Pagination } from '@elastic/eui/src/components/basic_table/pagination_bar';
import { render } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { NotificationItem } from '../../../../models/interfaces';
import { MOCK_DATA } from '../../../../test/mocks/mockData';
import { NotificationsTable } from '../components/NotificationsTable/NotificationsTable';
import { DEFAULT_PAGE_SIZE_OPTIONS } from '../utils/constants';

describe('<NotificationsTable /> spec', () => {
  configure({ adapter: new Adapter() });

  const sorting: EuiTableSortingType<NotificationItem> = {
    sort: {
      direction: 'asc',
      field: 'last_updated_time_ms',
    },
  };

  it('renders the component', () => {
  const pagination: Pagination = {
    pageIndex: 0,
    pageSize: 5,
    pageSizeOptions: DEFAULT_PAGE_SIZE_OPTIONS,
    totalItemCount: 0,
  };

    const utils = render(
      <NotificationsTable
        items={[]}
        onTableChange={() => {}}
        pagination={pagination}
        sorting={sorting}
        loading={false}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders the component with items', () => {
  const pagination: Pagination = {
    pageIndex: 0,
    pageSize: 5,
    pageSizeOptions: DEFAULT_PAGE_SIZE_OPTIONS,
    totalItemCount: 20,
  };

    const utils = render(
      <NotificationsTable
        items={MOCK_DATA.notifications.items}
        onTableChange={() => {}}
        pagination={pagination}
        sorting={sorting}
        loading={false}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });

});
