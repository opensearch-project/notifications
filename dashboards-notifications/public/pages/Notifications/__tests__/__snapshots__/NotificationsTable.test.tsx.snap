// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationsTable /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px 0px;"
  >
    <div
      class="euiFlexItem"
    >
      <h3
        class="euiTitle euiTitle--medium"
      >
        Notification history
      </h3>
    </div>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
  />
  <div
    style="padding: initial;"
  >
    <div
      class="euiBasicTable"
    >
      <div>
        <div
          class="euiTableHeaderMobile"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            />
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiTableSortMobile"
              >
                <div
                  class="euiPopover euiPopover--anchorDownRight"
                >
                  <div
                    class="euiPopover__anchor"
                  >
                    <button
                      class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                      type="button"
                    >
                      <span
                        class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                      >
                        <svg
                          aria-hidden="true"
                          class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                          focusable="false"
                          height="16"
                          role="img"
                          viewBox="0 0 16 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        />
                        <span
                          class="euiButtonEmpty__text"
                        >
                          Sorting
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <table
          class="euiTable euiTable--responsive euiTable--auto"
          id="random_html_id"
          tabindex="-1"
        >
          <caption
            class="euiScreenReaderOnly euiTableCaption"
          />
          <thead>
            <tr>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_event_source.title_0"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Notification"
                    >
                      Notification
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_event_source.feature_1"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Source type"
                    >
                      Source type
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_event_source.severity_2"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Severity"
                    >
                      Severity
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="ascending"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_last_updated_time_ms_3"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Time sent; Sorted in ascending order"
                    >
                      Time sent
                    </span>
                    <svg
                      aria-hidden="true"
                      aria-label="Sorted in ascending order"
                      class="euiIcon euiIcon--medium euiIcon-isLoading euiTableSortIcon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in descending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_success_4"
                role="columnheader"
                scope="col"
              >
                <div
                  class="euiTableCellContent"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Sent status"
                  >
                    Sent status
                  </span>
                </div>
              </th>
              <th
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_status_list_5"
                role="columnheader"
                scope="col"
              >
                <div
                  class="euiTableCellContent"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Channels"
                  >
                    Channels
                  </span>
                </div>
              </th>
              <th
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_status_list_6"
                role="columnheader"
                scope="col"
              >
                <div
                  class="euiTableCellContent"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Channel types"
                  >
                    Channel types
                  </span>
                </div>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="euiTableRow"
            >
              <td
                class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                colspan="7"
              >
                <div
                  class="euiTableCellContent euiTableCellContent--alignCenter"
                >
                  <span
                    class="euiTableCellContent__text"
                  >
                    No items found
                  </span>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`<NotificationsTable /> spec renders the component with items 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px 0px;"
  >
    <div
      class="euiFlexItem"
    >
      <h3
        class="euiTitle euiTitle--medium"
      >
        Notification history
      </h3>
    </div>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
  />
  <div
    style="padding: initial;"
  >
    <div
      class="euiBasicTable"
    >
      <div>
        <div
          class="euiTableHeaderMobile"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            />
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiTableSortMobile"
              >
                <div
                  class="euiPopover euiPopover--anchorDownRight"
                >
                  <div
                    class="euiPopover__anchor"
                  >
                    <button
                      class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                      type="button"
                    >
                      <span
                        class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                      >
                        <svg
                          aria-hidden="true"
                          class="euiIcon euiIcon--medium euiButtonContent__icon"
                          focusable="false"
                          height="16"
                          role="img"
                          viewBox="0 0 16 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                            fill-rule="non-zero"
                          />
                        </svg>
                        <span
                          class="euiButtonEmpty__text"
                        >
                          Sorting
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <table
          class="euiTable euiTable--responsive euiTable--auto"
          id="random_html_id"
          tabindex="-1"
        >
          <caption
            class="euiScreenReaderOnly euiTableCaption"
          />
          <thead>
            <tr>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_event_source.title_0"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Notification"
                    >
                      Notification
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_event_source.feature_1"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Source type"
                    >
                      Source type
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_event_source.severity_2"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Severity"
                    >
                      Severity
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="ascending"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_last_updated_time_ms_3"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Time sent; Sorted in ascending order"
                    >
                      Time sent
                    </span>
                    <svg
                      aria-label="Sorted in ascending order"
                      class="euiIcon euiIcon--medium euiTableSortIcon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 4.207v8.237c0 .307-.224.556-.5.556s-.5-.249-.5-.556V4.207L2.904 8.303a.5.5 0 01-.707-.707l4.242-4.242a1.5 1.5 0 012.122 0l4.242 4.242a.5.5 0 11-.707.707L8 4.207z"
                      />
                    </svg>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in descending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_success_4"
                role="columnheader"
                scope="col"
              >
                <div
                  class="euiTableCellContent"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Sent status"
                  >
                    Sent status
                  </span>
                </div>
              </th>
              <th
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_status_list_5"
                role="columnheader"
                scope="col"
              >
                <div
                  class="euiTableCellContent"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Channels"
                  >
                    Channels
                  </span>
                </div>
              </th>
              <th
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_status_list_6"
                role="columnheader"
                scope="col"
              >
                <div
                  class="euiTableCellContent"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Channel types"
                  >
                    Channel types
                  </span>
                </div>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="euiTableRow euiTableRow-isSelectable"
            >
              <td
                class="euiTableRowCell"
              >
                <div
                  class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                >
                  Notification
                </div>
                <div
                  class="euiTableCellContent euiTableCellContent--truncateText euiTableCellContent--overflowingContent"
                >
                  <button
                    class="euiLink euiLink--primary"
                    type="button"
                  >
                    test notification
                  </button>
                </div>
              </td>
              <td
                class="euiTableRowCell"
              >
                <div
                  class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                >
                  Source type
                </div>
                <div
                  class="euiTableCellContent euiTableCellContent--truncateText euiTableCellContent--overflowingContent"
                >
                  Reporting
                </div>
              </td>
              <td
                class="euiTableRowCell"
              >
                <div
                  class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                >
                  Severity
                </div>
                <div
                  class="euiTableCellContent euiTableCellContent--overflowingContent"
                >
                  Info
                </div>
              </td>
              <td
                class="euiTableRowCell"
              >
                <div
                  class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                >
                  Time sent
                </div>
                <div
                  class="euiTableCellContent euiTableCellContent--overflowingContent"
                >
                  06/07/21 4:53 pm GMT
                </div>
              </td>
              <td
                class="euiTableRowCell"
              >
                <div
                  class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                >
                  Sent status
                </div>
                <div
                  class="euiTableCellContent euiTableCellContent--overflowingContent"
                >
                  <div
                    class="euiHealth"
                  >
                    <div
                      class="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
                    >
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      >
                        <svg
                          aria-hidden="true"
                          class="euiIcon euiIcon--medium euiIcon--danger euiIcon-isLoading"
                          focusable="false"
                          height="16"
                          role="img"
                          viewBox="0 0 16 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        />
                      </div>
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      >
                        <button
                          class="euiLink euiLink--primary"
                          type="button"
                        >
                          Error
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </td>
              <td
                class="euiTableRowCell"
              >
                <div
                  class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                >
                  Channels
                </div>
                <div
                  class="euiTableCellContent euiTableCellContent--truncateText euiTableCellContent--overflowingContent"
                >
                  2 channels
                </div>
              </td>
              <td
                class="euiTableRowCell"
              >
                <div
                  class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                >
                  Channel types
                </div>
                <div
                  class="euiTableCellContent euiTableCellContent--overflowingContent"
                >
                  Slack, Email
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div>
        <div
          class="euiSpacer euiSpacer--m"
        />
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
        >
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiPopover euiPopover--anchorUpRight euiPopover--withTitle"
            >
              <div
                class="euiPopover__anchor"
              >
                <button
                  class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall"
                  data-test-subj="tablePaginationPopoverButton"
                  type="button"
                >
                  <span
                    class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiButtonContent__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                        fill-rule="non-zero"
                      />
                    </svg>
                    <span
                      class="euiButtonEmpty__text"
                    >
                      Rows per page
                      : 
                      5
                    </span>
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <nav
              class="euiPagination"
            >
              <button
                aria-label="Previous page"
                class="euiButtonIcon euiButtonIcon--text"
                data-test-subj="pagination-button-previous"
                disabled=""
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                  focusable="false"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </button>
              <ul
                class="euiPagination__list"
              >
                <li
                  class="euiPagination__item"
                >
                  <button
                    aria-controls="random_html_id"
                    aria-current="true"
                    aria-label="Page 1 of 4"
                    class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiButtonEmpty-isDisabled euiPaginationButton euiPaginationButton-isActive euiPaginationButton--hideOnMobile"
                    data-test-subj="pagination-button-0"
                    disabled=""
                    type="button"
                  >
                    <span
                      class="euiButtonContent euiButtonEmpty__content"
                    >
                      <span
                        class="euiButtonEmpty__text"
                      >
                        1
                      </span>
                    </span>
                  </button>
                </li>
                <li
                  class="euiPagination__item"
                >
                  <a
                    aria-controls="random_html_id"
                    aria-label="Page 2 of 4"
                    class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiPaginationButton euiPaginationButton--hideOnMobile"
                    data-test-subj="pagination-button-1"
                    href="#random_html_id"
                    rel="noreferrer"
                  >
                    <span
                      class="euiButtonContent euiButtonEmpty__content"
                    >
                      <span
                        class="euiButtonEmpty__text"
                      >
                        2
                      </span>
                    </span>
                  </a>
                </li>
                <li
                  class="euiPagination__item"
                >
                  <a
                    aria-controls="random_html_id"
                    aria-label="Page 3 of 4"
                    class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiPaginationButton euiPaginationButton--hideOnMobile"
                    data-test-subj="pagination-button-2"
                    href="#random_html_id"
                    rel="noreferrer"
                  >
                    <span
                      class="euiButtonContent euiButtonEmpty__content"
                    >
                      <span
                        class="euiButtonEmpty__text"
                      >
                        3
                      </span>
                    </span>
                  </a>
                </li>
                <li
                  class="euiPagination__item"
                >
                  <a
                    aria-controls="random_html_id"
                    aria-label="Page 4 of 4"
                    class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiPaginationButton euiPaginationButton--hideOnMobile"
                    data-test-subj="pagination-button-3"
                    href="#random_html_id"
                    rel="noreferrer"
                  >
                    <span
                      class="euiButtonContent euiButtonEmpty__content"
                    >
                      <span
                        class="euiButtonEmpty__text"
                      >
                        4
                      </span>
                    </span>
                  </a>
                </li>
              </ul>
              <a
                aria-controls="random_html_id"
                aria-label="Next page, 2"
                class="euiButtonIcon euiButtonIcon--text"
                data-test-subj="pagination-button-next"
                href="#random_html_id"
                rel="noreferrer"
              >
                <svg
                  aria-hidden="true"
                  class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                  focusable="false"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </a>
            </nav>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
